### PR TITLE
[flang][OpenMP] Use new modifiers in IF/LASTPRIVATE

### PR DESCRIPTION
**Pull request by kparzysz**
The usual changes, added more references to OpenMP specs.

### CODE REVIEW

File: flang/lib/Parser/openmp-parsers.cpp [hunk 1]
```diff
@@ -16,13 +16,58 @@
...
+  using NameWithId = std::pair<std::string, llvm::omp::Directive>;
+
+  llvm::iterator_range<const NameWithId *> directives() const;
+  void initTokens(NameWithId *) const;
+};
+
+llvm::iterator_range<const OmpDirectiveNameParser::NameWithId *>
+OmpDirectiveNameParser::directives() const {
+  static NameWithId table[llvm::omp::Directive_enumSize];
+  [[maybe_unused]] static bool init = (initTokens(table), true);
```
Comment (tblah) [inline flang/lib/Parser/openmp-parsers.cpp:54]: If I understand correctly, this is a trick to run `initTokens` along with static variable initializers.

This is quite clever but I had to stare at it for a while to understand what was going on. I would prefer something very simple like, but I will leave this up to you - maybe this is a widely understood idiom I am just unfamiliar with.
```c++
static bool init{false};
if (!init) {
  init_tokens(table);
  init = true;
}
```
This isn't semantically identical but I don't think running initTokens a bit later should matter.

**Review by tblah — APPROVED** [review]
LGTM, just some nit picks.

File: flang/lib/Parser/openmp-parsers.cpp [hunk 2]
```diff
@@ -16,13 +16,58 @@
...
 
 // OpenMP Directives and Clauses
 namespace Fortran::parser {
 
 constexpr auto startOmpLine = skipStuffBeforeStatement >> "!$OMP "_sptok;
 constexpr auto endOmpLine = space >> endOfLine;
 
+/// Parse OpenMP directive name (this includes compound directives).
+struct OmpDirectiveNameParser {
+  using resultType = llvm::omp::Directive;
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Parser/openmp-parsers.cpp:32]: This spelling is actually mandated by the parsing infrastructure.  See this comment: https://github.com/llvm/llvm-project/blob/main/flang/lib/Parser/basic-parsers.h#L12-L20

File: flang/lib/Parser/openmp-parsers.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Parser/openmp-parsers.cpp:54]: The issue here is with thread-safety.  Starting with C++11, the standard guarantees that local statics are initialized in a thread-safe manner, that is, that the initializer code is only executed by a single thread.  Hence the "trick" is to make the preparation of a local static be a part of the initializer.  This is a de facto standard in the LLVM code, although I didn't see it mentioned in the coding guidelines.  It's somewhat rare, but within the set of use cases it's the preferred solution.

**Pull request merged**